shadow/state.go

```diff
@@ -0,0 +1,117 @@
+package shadow
+
+import (
+	"encoding/json"
+	"errors"
+)
+
+// ErrVersionConflict means thing state update was aborted due to version conflict.
+var ErrVersionConflict = errors.New("version conflict")
+
+// ThingState represents Thing Shadow State.
+type ThingState struct {
+	Desired  map[string]interface{} `json:"desired,omitempty"`
+	Reported map[string]interface{} `json:"reported,omitempty"`
+	Delta    map[string]interface{} `json:"delta,omitempty"`
+}
+
+// ThingDocument represents Thing Shadow Document.
+type ThingDocument struct {
+	State     ThingState `json:"state"`
+	Version   int        `json:"version,omitempty"`
+	Timestamp int        `json:"timestamp,omitempty"`
+}
+
+type thingStateRaw struct {
+	Desired  json.RawMessage `json:"desired,omitempty"`
+	Reported json.RawMessage `json:"reported,omitempty"`
+	Delta    json.RawMessage `json:"delta,omitempty"`
+}
+
+type thingDocumentRaw struct {
+	State     thingStateRaw `json:"state"`
+	Version   int           `json:"version,omitempty"`
+	Timestamp int           `json:"timestamp,omitempty"`
+}
+
+type thingDelta struct {
+	State     map[string]interface{} `json:"state"`
+	Version   int                    `json:"version,omitempty"`
+	Timestamp int                    `json:"timestamp,omitempty"`
+}
+
+func (s *ThingDocument) update(state *thingDocumentRaw) error {
+	if s.Version > state.Version {
+		// Received an old version; just ignore it.
+		return nil
+	}
+	s.Version = state.Version
+	s.Timestamp = state.Timestamp
+	if err := updateStateRaw(s.State.Desired, state.State.Desired); err != nil {
+		return err
+	}
+	if err := updateStateRaw(s.State.Reported, state.State.Reported); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *ThingDocument) updateDelta(state *thingDelta) (bool, error) {
```

Comment:
Does this function need to return (bool, error)? (There seems no path for returning err != nil

Reply:
Removed error from return values.